src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -179,52 +179,63 @@ public async Task InitializeAsync()
-                }
-            }
+            Block block = await this.GetBlockAsync(blockId).ConfigureAwait(false);
 
-            return this.blockRepository.GetTransactionByIdAsync(trxid);
+            return block?.Transactions.FirstOrDefault(x => x.GetHash() == trxid);
         }
 
         /// <inheritdoc/>
-        public Task<uint256> GetBlockIdByTransactionIdAsync(uint256 trxid)
+        public async Task<uint256> GetBlockIdByTransactionIdAsync(uint256 trxid)
         {
             lock (this.blocksCacheLock)
             {
-                foreach (ChainedHeaderBlock chainedHeaderBlock in this.pendingBlocksCache.Values)
+                // According to SaveBlockAsync the last block in the list is considered to be on the current main chain.
+                if (this.chain.GetBlock(this.BlockStoreCacheTip.HashBlock) != null)
                 {
-                    bool exists = chainedHeaderBlock.Block.Transactions.Any(x => x.GetHash() == trxid);
-
-                    if (exists)
+                    // Indeed so let's proceed.
+                    for (ChainedHeader chainedHeader = this.BlockStoreCacheTip; chainedHeader != null; chainedHeader = chainedHeader.Previous)
                     {
-                        uint256 blockId = chainedHeaderBlock.Block.GetHash();
+                        if (!this.pendingBlocksCache.TryGetValue(chainedHeader.HashBlock, out ChainedHeaderBlock chainedHeaderBlock))
+                            break;
 
-                        this.logger.LogTrace("Block Id '{0}' with tx '{1}' was found in the pending blocks cache.", blockId, trxid);
-                        return Task.FromResult(blockId);
+                        bool exists = chainedHeaderBlock.Block.Transactions.Any(x => x.GetHash() == trxid);
+
+                        if (exists)
+                        {
+                            uint256 blockId = chainedHeaderBlock.Block.GetHash();
+
+                            this.logger.LogTrace("Block Id '{0}' with tx '{1}' was found in the pending blocks cache.", blockId, trxid);
+                            return blockId;
+                        }
                     }
                 }
             }
 
-            return this.blockRepository.GetBlockIdByTransactionIdAsync(trxid);
+            // Now look in the repository.
+            uint256 blockId2 = await this.blockRepository.GetBlockIdByTransactionIdAsync(trxid).ConfigureAwait(false);
+
+            // The repository may be dirty. Check that this is a valid block.
+            if (blockId2 == null || this.chain.GetBlock(blockId2) == null)
+                return null;
+
+            return blockId2;
         }
 
         /// <inheritdoc/>
         public async Task<Block> GetBlockAsync(uint256 blockHash)
         {
             lock (this.blocksCacheLock)
             {
+                // The cache or repository may be dirty. Check that this is a valid block.
+                if (this.chain.GetBlock(blockHash) == null)
```

Comment:
There is not really a need for this check anything that is in `chain` will be in `pendingBlocksCache`minus a tiny race of the time it took to update `chain` and to push to `pendingBlocksCache` but peers asking for this block